File: AZQrCodeScanController/AZQrCodeScanController/AZQrCodeScanController.m
//
//  AZQrCodeScanController.m
//  AZQrCodeScanController
//
//  Created by tianfengyu on 2017/6/9.
//  Copyright © 2017年 Azreal. All rights reserved.
//

#import "AZQrCodeScanController.h"
#import "AZQrCodeScanView.h"
#import "AZQrCodeScanDevice.h"

#define AZ_SCREENWIDTH [UIScreen mainScreen].bounds.size.width
#define AZ_SCREENHEIGHT [UIScreen mainScreen].bounds.size.height
#define SCANPADDING 50

@interface AZQrCodeScanController ()

@property (nonatomic, assign) CGRect scanFrame;
@property (nonatomic, strong) AZQrCodeScanDevice *device;
@property (nonatomic, strong) AZQrCodeScanView *scanView;
@property (nonatomic, copy) void(^scanCompleteBlock)(NSString *result);
@property (nonatomic, strong) UINavigationBar *navigationBar;
/// 记录进入该控制器前的导航栏状态
@property (nonatomic, strong) NSDictionary *originalTitleTextAttributes;
@property (nonatomic, strong) UIColor *originalTintColor;
@property (nonatomic, strong) UIImage *originalShadowImage;
@property (nonatomic, strong) UIImage *originalBackgroundImage;

@end

@implementation AZQrCodeScanController

// MARK: - 初始化方法
- (instancetype)initWithScanComplete:(void (^)(NSString *, AZQrCodeScanController *))complete {
    
    if (self = [super init]) {
        
        [self configInitValue];
        
        CGFloat width = AZ_SCREENWIDTH - SCANPADDING*2;
        _scanFrame = CGRectMake((AZ_SCREENWIDTH-width)/2,
                                (AZ_SCREENHEIGHT-width)/2,
                                width,
                                width);
        _scanView = [[AZQrCodeScanView alloc] initWithScanFrame:_scanFrame];
        __weak AZQrCodeScanController *wkSelf = self;
        _scanCompleteBlock = ^(NSString *result) {
            wkSelf.scanView.timerState = AZTimerStateStop;
            complete(result, wkSelf);
        };
    }
    return self;
}

- (instancetype)initWithScanFrame:(CGRect)frame complete:(void (^)(NSString *, AZQrCodeScanController *))complete {
    
    if (self = [super init]) {
        [self configInitValue];
        _scanFrame = frame;
        _scanView = [[AZQrCodeScanView alloc] initWithScanFrame:_scanFrame];
        __weak AZQrCodeScanController *wkSelf = self;
        _scanCompleteBlock = ^(NSString *result) {
            wkSelf.scanView.timerState = AZTimerStateStop;
            complete(result, wkSelf);
        };
    }
    return self;
}

- (void)configInitValue {
    _navigationTintColor = [UIColor whiteColor];
    _navigationBarAlpha = 0;
    _navigationBarTintColor = [UIColor whiteColor];
    _navigationTitleText = @"二维码扫描";
}

/**
 启动扫码线动画与session
 */
- (void)start {
    _scanView.timerState = AZTimerStateMove;
    if (![_device.session isRunning]) {
        [_device.session startRunning];
    }
}

/**
 暂停扫码线动画与session
 */
- (void)stop {
    _scanView.timerState = AZTimerStateStop;
    if ([_device.session isRunning]) {
        [_device.session stopRunning];
    }
}

// MARK: - 生命周期
- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.whiteColor;
    [self.view addSubview:_scanView];
    [self requestCaptureAuth];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self configNavigation];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self start];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self stop];
    
    self.navigationController.navigationBar.shadowImage = _originalShadowImage;
    self.navigationController.navigationBar.titleTextAttributes = _originalTitleTextAttributes;
    self.navigationController.navigationBar.tintColor = _originalTintColor;
    [self.navigationController.navigationBar setBackgroundImage:_originalBackgroundImage forBarMetrics:UIBarMetricsDefault];
}



- (void)dealloc {
    if (_device && _device.session.isRunning) {
        [_device.session stopRunning];
    }
    dispatch_source_cancel(_scanView.timer);
}

- (void)configNavigation {
    if (self.navigationController == nil) {
        _navigationBar = [[UINavigationBar alloc] initWithFrame:CGRectMake(0, 0, AZ_SCREENWIDTH, 64)];
        _navigationBar.shadowImage = [[UIImage alloc] init];
        [_navigationBar setBackgroundImage:[self imageWithColor:_navigationBarTintColor alpha:_navigationBarAlpha] forBarMetrics:UIBarMetricsDefault];
        _navigationBar.tintColor = _navigationTintColor;
        UINavigationItem *navigationItem = [[UINavigationItem alloc] initWithTitle:_navigationTitleText];
        [_navigationBar pushNavigationItem:navigationItem animated:true];
        NSString *bundlePath = [[NSBundle bundleForClass:[self class]] pathForResource:@"AZQrCode" ofType:@"bundle"];
        NSString *imagePath = [[NSBundle bundleWithPath:bundlePath] pathForResource:@"close@2x" ofType:@"png"];
        UIImage *image = [[UIImage imageWithContentsOfFile:imagePath] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        UIBarButtonItem *closeItem = [[UIBarButtonItem alloc] initWithImage:image style:UIBarButtonItemStylePlain target:self action:@selector(dismissController)];
        navigationItem.leftBarButtonItem = closeItem;
        _navigationBar.titleTextAttributes = @{NSForegroundColorAttributeName: _navigationTintColor};
//        UIBarButtonItem *albumItem = [[UIBarButtonItem alloc] initWithTitle:@"相册" style:UIBarButtonItemStylePlain target:self action:@selector(albumClick)];
//        navigationItem.rightBarButtonItem = albumItem;
        [self.view addSubview:_navigationBar];
    } else {
        self.navigationItem.title = _navigationTitleText;
        
        self.originalTitleTextAttributes = self.navigationController.navigationBar.titleTextAttributes;
        self.originalTintColor = self.navigationController.navigationBar.tintColor;
        self.originalShadowImage = self.navigationController.navigationBar.shadowImage;
        self.originalShadowImage = [self.navigationController.navigationBar backgroundImageForBarMetrics:UIBarMetricsDefault];
        
        self.navigationController.navigationBar.titleTextAttributes = @{NSForegroundColorAttributeName: _navigationTintColor};
//        UIBarButtonItem *albumItem = [[UIBarButtonItem alloc] initWithTitle:@"相册" style:UIBarButtonItemStylePlain target:self action:@selector(albumClick)];
//        self.navigationItem.rightBarButtonItem = albumItem;
        self.navigationController.navigationBar.tintColor = _navigationTintColor;
        self.navigationController.navigationBar.shadowImage = [[UIImage alloc] init];
        [self.navigationController.navigationBar setBackgroundImage:[self imageWithColor:_navigationBarTintColor alpha:_navigationBarAlpha] forBarMetrics:UIBarMetricsDefault];
    }
}

- (UIImage *)imageWithColor:(UIColor *)color alpha:(CGFloat)alpha {
    CGRect rect = CGRectMake(0, 0, 1, 1);
    UIGraphicsBeginImageContext(rect.size);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetFillColorWithColor(context, color.CGColor);
    CGContextSetAlpha(context, alpha);
    CGContextFillRect(context, rect);
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return image;
}

- (void)albumClick {
    
}

- (void)dismissController {
    [self dismissViewControllerAnimated:true completion:nil];
}

- (void)requestCaptureAuth {
    __weak AZQrCodeScanController *wkSelf = self;
    
    AVAuthorizationStatus state = [AVCaptureDevice authorizationStatusForMediaType:AVMediaTypeVideo];
    switch (state) {
        case AVAuthorizationStatusNotDetermined: {
            // 用户还没有决定是否给相机授权
            [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
                dispatch_async(dispatch_get_main_queue(), ^{
                    if (granted) {
                        // 用户接受
                        [wkSelf deviceWork];
                    } else {
                        // 用户拒绝
                        [wkSelf showPrompt:nil];
                    }
                });
            }];
        }
            break;
        case AVAuthorizationStatusDenied:
        case AVAuthorizationStatusRestricted:
            // 用户拒绝相机授权或没有相机权限
            [self showPrompt:nil];
            break;
        case AVAuthorizationStatusAuthorized:
            // 授权
            [self deviceWork];
            break;
    }
}

- (void)deviceWork {
    if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
        _device = [[AZQrCodeScanDevice alloc] initWithScanFrame:_scanFrame layer:self.view.layer];
        _device.complete = _scanCompleteBlock;
    } else {
        [self showPrompt:@"当前设备没有拍照功能"];
    }
    
}

- (void)showPrompt:(NSString *)text {
    
    _scanView.hidden = true;
    
    UILabel *promptView = [[UILabel alloc] initWithFrame:CGRectMake(20, 0, AZ_SCREENWIDTH-40, 300)];
    promptView.textAlignment = NSTextAlignmentCenter;
    
    NSString *promptText = nil;
    if (text == nil) {
        if (!_appName) {
            _appName = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleDisplayName"] ?
            [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleDisplayName"] :
            [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleName"];
        }
        promptText = [NSString stringWithFormat:@"请在iPhone的\"设置-隐私-相机\"中允许%@访问您的相机", _appName];
    } else {
        promptText = text;
    }
    promptView.text = promptText;
    promptView.numberOfLines = 0;
    [self.view addSubview:promptView];
    
}

// MARK: - Set
- (void)setScanLineImage:(UIImage *)scanLineImage {
    _scanLineImage = scanLineImage;
    _scanView.scanLine.image = scanLineImage;
}

- (void)setScanImage:(UIImage *)scanImage {
    _scanImage = scanImage;
    _scanView.scanImageView.image = scanImage;
}

- (void)setTintColor:(UIColor *)tintColor {
    _tintColor = tintColor;
    _scanView.scanImageView.tintColor = tintColor;
    _scanView.scanLine.tintColor = tintColor;
}

- (void)setScanColor:(UIColor *)scanColor {
    _scanColor = scanColor;
    _scanView.scanImageView.tintColor = scanColor;
}

- (void)setScanLineColor:(UIColor *)scanLineColor {
    _scanLineColor = scanLineColor;
    _scanView.scanLine.tintColor = scanLineColor;
}

- (void)setCoverViewAlpha:(CGFloat)coverViewAlpha {
    _coverViewAlpha = coverViewAlpha;
    for (UIView *item in @[_scanView.topCoverView, _scanView.leftCoverView, _scanView.bottomCoverView, _scanView.rightCoverView]) {
        item.alpha = coverViewAlpha;
    }
}

- (void)setIntroduceText:(NSString *)introduceText {
    _introduceText = introduceText;
    _scanView.introduceLabel.text = introduceText;
}

- (void)setIntroduceFontSize:(CGFloat)introduceFontSize {
    _introduceFontSize = introduceFontSize;
    _scanView.introduceLabel.font = [UIFont systemFontOfSize:introduceFontSize];
    [_scanView.introduceLabel sizeToFit];
}

- (void)setIntroduceFont:(UIFont *)introduceFont {
    _introduceFont = introduceFont;
    _scanView.introduceLabel.font = introduceFont;
    [_scanView.introduceLabel sizeToFit];
}

- (void)setIntroduceTextColor:(UIColor *)introduceTextColor {
    _introduceTextColor = introduceTextColor;
    _scanView.introduceLabel.textColor = introduceTextColor;
}

- (void)setIntroduceFrame:(CGRect)introduceFrame {
    _introduceFrame = introduceFrame;
    _scanView.introduceLabel.frame = introduceFrame;
}

@end

File: AZQrCodeScanController/AZQrCodeScanController-Swift/AZQrCodeScanController.swift
//
//  AZQrCodeCapture.swift
//  AZQrCodeCapture
//
//  Created by tianfengyu on 2017/5/21.
//  Copyright © 2017年 tianfengyu. All rights reserved.
//

import UIKit
import AVFoundation

private let AZ_screenWidth = UIScreen.main.bounds.width
private let AZ_screenHeight = UIScreen.main.bounds.height
private let scanImageLeftPadding: CGFloat = 50.0

class AZQrCodeScanController: UIViewController {
    
    /// 扫码线图片
    var scanLineImage: UIImage? { didSet { scanView.scanLine.image = scanLineImage } }
    
    /// 扫码框图片
    var scanImage: UIImage? { didSet { scanView.scanImageView.image = scanImage } }
    
    /// 扫码框和扫码线颜色
    var tintColor: UIColor! { didSet {
        scanView.scanImageView.tintColor = scanColor
        scanView.scanLine.tintColor = scanColor
        } }
    
    /// 单独设置扫码框颜色
    var scanColor: UIColor! { didSet { scanView.scanImageView.tintColor = scanColor } }
    
    /// 单独设置扫码线颜色
    var scanLineColor: UIColor! { didSet { scanView.scanLine.tintColor = scanLineColor } }
    
    /// 遮罩层透明度
    var coverViewAlpha: CGFloat! { didSet {
        for item in [scanView.topCoverView, scanView.leftCoverView, scanView.bottomCoverView, scanView.rightCoverView] {
            item.alpha = coverViewAlpha
        } } }
    
    /// 提示文字
    var introduceText: String? { didSet { scanView.introduceLabel.text = introduceText } }
    
    /// 提示文字字体大小
    var introduceFontSize: CGFloat! { didSet {
        scanView.introduceLabel.font = UIFont.systemFont(ofSize: introduceFontSize)
        scanView.introduceLabel.sizeToFit()
        } }
    
    /// 提示文字字体
    var introduceFont: UIFont! { didSet {
        scanView.introduceLabel.font = introduceFont
        scanView.introduceLabel.sizeToFit()
        } }
    
    /// 提示文字颜色
    var introduceTextColor: UIColor! { didSet { scanView.introduceLabel.textColor = introduceTextColor } }
    
    /// 提示文字位置
    var introduceFrame: CGRect! { didSet { scanView.introduceLabel.frame = introduceFrame } }
    
    
    /// 扫码框位置
    private var scanFrame: CGRect!
    
    /// 扫码完成回调
    private var complete: ((String)->())?
    
    private var device: AZQrCodeScanDevice?
    private var scanView: AZQrCodeScanView!
    
    public convenience init(complete: ((String)->())?) {
        self.init(nibName: nil, bundle: nil)
        let width = AZ_screenWidth-scanImageLeftPadding*2
        self.scanFrame = CGRect(x: (AZ_screenWidth-width)/2,
                                y: (AZ_screenHeight-width)/2,
                                width: width,
                                height: width)
        scanView = AZQrCodeScanView(scanFrame: self.scanFrame)
        self.complete = complete
    }
    
    public convenience init(scanFrame: CGRect, complete: ((String)->())?) {
        self.init(nibName: nil, bundle: nil)
        self.scanFrame = scanFrame
        scanView = AZQrCodeScanView(scanFrame: self.scanFrame)
        self.complete = complete
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.addSubview(scanView)
        device = AZQrCodeScanDevice(scanFrame: scanFrame, layer: view.layer)
        device?.complete = complete
    }
    
    private override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    deinit {
        if device != nil && device!.session.isRunning {
            device!.session.stopRunning()
        }
        if !scanView.timer.isCancelled {
            scanView.timer.cancel()
        }
    }
}

private class AZQrCodeScanView: UIView {
    var topCoverView = UIView()
    var leftCoverView = UIView()
    var rightCoverView = UIView()
    var bottomCoverView = UIView()
    
    fileprivate var scanFrame: CGRect!
    
    var scanImageView: UIImageView! // 扫码框
    var scanLine: UIImageView! // 扫码线
    var introduceLabel: UILabel! // 提示文字label
    
    
    /// 扫码线移动方向
    private enum LineMoveDirect {
        case up
        case down
    }
    var timer: DispatchSourceTimer!
    private var lineDirection = LineMoveDirect.down
    
    convenience init(scanFrame: CGRect) {
        self.init(frame: CGRect(x: 0, y: 0, width: AZ_screenWidth, height: AZ_screenHeight))
        configCoverView()
        
        self.scanFrame = scanFrame
        resetCoverViewFrame()
        configScanUI()
        configTimer()
    }
    
    /// 扫码线移动
    private func configTimer() {
        timer = DispatchSource.makeTimerSource(flags: .strict, queue: .main)
        timer.scheduleRepeating(deadline: .now(), interval: .milliseconds(10))
        weak var wkSelf = self
        timer.setEventHandler {
            var lineFrame = wkSelf!.scanLine.frame
            switch wkSelf!.lineDirection {
            case .up:
                lineFrame.origin.y -= 1
            case .down:
                lineFrame.origin.y += 1
            }
            wkSelf?.scanLine.frame = lineFrame
            if lineFrame.origin.y >= wkSelf!.scanFrame.origin.y + wkSelf!.scanFrame.width - lineFrame.size.height {
                wkSelf!.lineDirection = .up
            } else if lineFrame.origin.y <= wkSelf!.scanFrame.origin.y {
                wkSelf!.lineDirection = .down
            }
        }
        timer.resume()
    }
    
    /// 扫码框/扫码线/介绍文字
    private func configScanUI() {
        
        let bundlePath = Bundle.main.path(forResource: "AZQrCode", ofType: "bundle")!
        let captureBundle = Bundle(path: bundlePath)!
        
        scanImageView = UIImageView(frame: scanFrame)
        let bgPath = captureBundle.path(forResource: "scan_bg_pic@2x", ofType: "png")!
        scanImageView.image = UIImage(contentsOfFile: bgPath)
        addSubview(scanImageView)
        
        scanLine = UIImageView(frame: CGRect(x: scanFrame.origin.x, y: scanFrame.origin.y
            , width: scanFrame.width, height: 2))
        let linePath = captureBundle.path(forResource: "scan_line@2x", ofType: "png")!
        scanLine.image = UIImage(contentsOfFile: linePath)
        addSubview(scanLine)
        
        introduceLabel = UILabel(frame: CGRect(x: scanFrame.origin.x, y: scanFrame.origin.y+scanFrame.size.height+20, width: scanFrame.width, height: 40))
        introduceLabel.numberOfLines = 0
        introduceLabel.textAlignment = .center
        introduceLabel.text = "将二维码/条码放入框内，即可自动扫描。"
        introduceLabel.textColor = .white
        introduceLabel.font = UIFont.systemFont(ofSize: 14)
        addSubview(introduceLabel)
    }
    
    /// 配置coverView
    private func configCoverView() {
        for item in [topCoverView, leftCoverView, bottomCoverView, rightCoverView] {
            item.backgroundColor = UIColor(white: 0, alpha: 0.4)
            addSubview(item)
        }
    }
    
    /// 设置遮罩层位置
    private func resetCoverViewFrame() {
        leftCoverView.frame = CGRect(x: 0, y: 0, width: scanFrame.origin.x, height: AZ_screenHeight)
        topCoverView.frame = CGRect(x: scanFrame.origin.x, y: 0, width: AZ_screenWidth, height: scanFrame.origin.y)
        rightCoverView.frame = CGRect(x: scanFrame.origin.x, y: scanFrame.height+scanFrame.origin.y, width: AZ_screenWidth, height: AZ_screenHeight)
        bottomCoverView.frame = CGRect(x: scanFrame.width+scanFrame.origin.x, y: scanFrame.origin.y, width: AZ_screenWidth, height: scanFrame.height)
    }
    
    
    override init(frame: CGRect) { super.init(frame: frame) }
    required init?(coder aDecoder: NSCoder) { fatalError("init(coder:) has not been implemented") }
}

private class AZQrCodeScanDevice: NSObject, AVCaptureMetadataOutputObjectsDelegate {
    
    private var device: AVCaptureDevice!
    private var input: AVCaptureDeviceInput!
    private var output: AVCaptureMetadataOutput!
    var session: AVCaptureSession!
    private var preview: AVCaptureVideoPreviewLayer!
    var complete: ((String)->())?
    
    init(scanFrame: CGRect, layer: CALayer) {
        super.init()
        device = AVCaptureDevice.defaultDevice(withMediaType: AVMediaTypeVideo)
        do {
            input = try AVCaptureDeviceInput(device: device)
        } catch {
            print(error)
        }
        output = AVCaptureMetadataOutput()
        output.setMetadataObjectsDelegate(self, queue: .main)
        
        session = AVCaptureSession()
        session.sessionPreset = AVCaptureSessionPresetHigh
        if session.canAddInput(input) {
            session.addInput(input)
        }
        if session.canAddOutput(output) {
            session.addOutput(output)
        }
        output.metadataObjectTypes = [AVMetadataObjectTypeQRCode, AVMetadataObjectTypeEAN13Code, AVMetadataObjectTypeEAN8Code, AVMetadataObjectTypeCode128Code]
        output.rectOfInterest = CGRect(x: scanFrame.origin.y/AZ_screenHeight,
                                       y: (AZ_screenWidth-scanFrame.size.width-scanFrame.origin.x)/AZ_screenWidth,
                                       width: scanFrame.size.height/AZ_screenHeight,
                                       height: scanFrame.size.width/AZ_screenWidth)
        preview = AVCaptureVideoPreviewLayer(session: session)
        preview.videoGravity = AVLayerVideoGravityResizeAspectFill
        preview.frame = CGRect(x: 0, y: 0, width: AZ_screenWidth, height: AZ_screenHeight)
        layer.insertSublayer(preview, at: 0)
        session.startRunning()
    }
    
    func captureOutput(_ captureOutput: AVCaptureOutput!, didOutputMetadataObjects metadataObjects: [Any]!, from connection: AVCaptureConnection!) {
        if metadataObjects.count > 0 {
            session.stopRunning()
            let metadataObject = metadataObjects[0] as! AVMetadataMachineReadableCodeObject
            let stringValue = metadataObject.stringValue
            guard stringValue != nil else {
                complete?("")
                return
            }
            complete?(stringValue!)
        }
    }
}
